include/led.h

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (c) 2017 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef _ZEPHYR_LED_H
+#define _ZEPHYR_LED_H
+
+/**
+ * @file
+ * @brief Public LED driver APIs
+ */
+
+#include <zephyr/types.h>
+#include <device.h>
+
+/**
+ * @brief Driver specific representation of a LED device
+ */
+
+struct led_data {
+	/* Minimum acceptable time period (in ms) supported by the device */
+	u32_t min_period;
+	/* Maximum acceptable time period (in ms) supported by the device */
+	u32_t max_period;
+	/* Minimum acceptable brightness (in percent) supported by the device */
+	u32_t min_brightness;
+	/* Maximum acceptable brightness (in percent) supported by the device */
+	u32_t max_brightness;
+};
+
+/**
+ * @typedef led_api_blink()
+ * @brief Callback API for blinking a LED
+ *
+ * @see led_blink() for argument descriptions.
+ */
+typedef int (*led_api_blink)(struct device *dev, u32_t led, u32_t period);
+
+/**
+ * @typedef led_api_blink_oneshot()
+ * @brief Callback API for oneshot blink of a LED
+ *
+ * @see led_blink_oneshot() for argument descriptions.
+ */
+typedef int (*led_api_blink_oneshot)(struct device *dev, u32_t led);
+
+/**
+ * @typedef led_api_set_brightness()
+ * @brief Callback API for setting brightness of a LED
+ *
+ * @see led_set_brightness() for argument descriptions.
+ */
+typedef int (*led_api_set_brightness)(struct device *dev, u32_t led,
+				      u32_t value);
+/**
+ * @typedef led_api_on()
+ * @brief Callback API for turning on an LED
+ *
+ * @see led_on() for argument descriptions.
+ */
+typedef int (*led_api_on)(struct device *dev, u32_t led);
```

Comment:
Above, I proposed specifying a standard range of brightness that all drivers need to support (even if they ignore low order bits).

If that's done, then `led_on` and `led_off` can be implemented as static inlines that delegate to `led_set_brightness`:

```
static inline int led_on(struct device *dev, u32_t led)
{
	return led_set_brightness(dev, led, MAX_BRIGHTNESS);
}

static inline int led_off(struct device *dev, u32_t led)
{
	return led_set_brightness(dev, led, MIN_BRIGHTNESS);
}
```

This will save code space without restricting the API.

Reply:
Agree, the `led_set_brightness` API will be in following structure:
```
int led_set_brightness(struct device *dev, u32_t led, u8_t value);
```

Reply:
I have clarified this before during early conversation with @mbolivar. Even though we can turn off an LED by setting brightness to MIN_BRIGHTNESS, still the led would be on. The device would be consuming some current. We need to explicitly turn off the LED's and that's the whole purpose of having `led_on` and `led_off` API's.

If there are devices capable of completely turning off an LED by setting MIN_BRIGHTNESS they can ignore the above API's and implement the `led_set_brightness` alone.